### ISSUE TITLE
Current swap counter overflow (?)

### ISSUE THREAD

**Issue by zerkms** [lxc/lxd#3482]
# Required information

 * Distribution: ubuntu
 * Distribution version: xenial
 * The output of "lxc info" or if that fails:
```
config: {}
api_extensions:
- id_map
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE-----
  certificate_fingerprint: ...
  driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-83-generic
  server: lxd
  server_pid: 1466
  server_version: 2.0.10
  storage: dir
  storage_version: ""
```

# Issue description

Sometimes `lxc info <container-name>` command returns the negative value of swap used (and there is no the peak swap usage there at all).

# Steps to reproduce

 1. Run a container with low memory limit (eg 100Mb)
 2. Run any tool/command that consumes memory close to the limit, eg `stress --vm-bytes 90M -m 1`
 3. On the host run `lxc info <container-name>` until you're lucky

# Information to attach

An exact unmodified output of the `lxc info second` I've just got:
```
Name: second
Remote: unix:/var/lib/lxd/unix.socket
Architecture: x86_64
Created: 2017/06/30 03:37 UTC
Status: Running
Type: persistent
Profiles: default
Pid: 2028
Ips:
  eth0: inet    10.50.200.19    vethVFRKLF
  eth0: inet6   fe80::216:3eff:fe63:c30b        vethVFRKLF
  lo:   inet    127.0.0.1
  lo:   inet6   ::1
Resources:
  Processes: 18
  Memory usage:
    Memory (current): 97.58MB
    Memory (peak): 100.00MB
    Swap (current): -102318081B
  Network usage:
    eth0:
      Bytes received: 258.49kB
      Bytes sent: 5.09kB
      Packets received: 1238
      Packets sent: 56
    lo:
      Bytes received: 9.08kB
      Bytes sent: 9.08kB
      Packets received: 120
      Packets sent: 120
```

**Comment by stgraber** [comment]
Can you run the same "lxc info" with "--debug" so we can see the raw API reply too?

**Comment by zerkms** [comment]
```
# lxc info second --debug
DBUG[07-02|21:54:11] Raw response: {"type":"sync","status":"Success","status_code":200,"operation":"","error_code":0,"error":"","metadata":{"config":{},"api_extensions":["id_map"],"api_status":"stable","api_version":"1.0","auth":"trusted","public":false,"environment":{"addresses":[],"architectures":["x86_64","i686"],"certificate":"-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----\n","certificate_fingerprint":"...","driver":"lxc","driver_version":"2.0.8","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"4.4.0-83-generic","server":"lxd","server_pid":1466,"server_version":"2.0.10","storage":"dir","storage_version":""}}}

DBUG[07-02|21:54:11] Raw response: {"type":"sync","status":"Success","status_code":200,"operation":"","error_code":0,"error":"","metadata":{"architecture":"x86_64","config":{"boot.autostart":"1","limits.cpu":"1","limits.memory":"100MB","volatile.base_image":"ac23a9d0febb30d981d600e73aaa754e53affc1475abda6b59a31b771e1164aa","volatile.eth0.hwaddr":"00:16:3e:63:c3:0b","volatile.idmap.base":"0","volatile.idmap.next":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]","volatile.last_state.power":"RUNNING"},"devices":{"root":{"path":"/","type":"disk"}},"ephemeral":false,"profiles":["default"],"stateful":false,"created_at":"2017-06-30T03:37:39Z","expanded_config":{"boot.autostart":"1","limits.cpu":"1","limits.memory":"100MB","volatile.base_image":"ac23a9d0febb30d981d600e73aaa754e53affc1475abda6b59a31b771e1164aa","volatile.eth0.hwaddr":"00:16:3e:63:c3:0b","volatile.idmap.base":"0","volatile.idmap.next":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":100000,\"Nsid\":0,\"Maprange\":65536}]","volatile.last_state.power":"RUNNING"},"expanded_devices":{"eth0":{"name":"eth0","nictype":"bridged","parent":"br0","type":"nic"},"root":{"path":"/","type":"disk"}},"name":"second","status":"Running","status_code":103}}

DBUG[07-02|21:54:11] Raw response: {"type":"sync","status":"Success","status_code":200,"operation":"","error_code":0,"error":"","metadata":{"status":"Running","status_code":103,"disk":{},"memory":{"usage":102203392,"usage_peak":104857600,"swap_usage":-90963968,"swap_usage_peak":0},"network":{"eth0":{"addresses":[{"family":"inet","address":"10.50.200.19","netmask":"16","scope":"global"},{"family":"inet6","address":"fe80::216:3eff:fe63:c30b","netmask":"64","scope":"link"}],"counters":{"bytes_received":300717,"bytes_sent":5217,"packets_received":1708,"packets_sent":56},"hwaddr":"00:16:3e:63:c3:0b","host_name":"vethVFRKLF","mtu":1500,"state":"up","type":"broadcast"},"lo":{"addresses":[{"family":"inet","address":"127.0.0.1","netmask":"8","scope":"local"},{"family":"inet6","address":"::1","netmask":"128","scope":"local"}],"counters":{"bytes_received":9296,"bytes_sent":9296,"packets_received":120,"packets_sent":120},"hwaddr":"","host_name":"","mtu":65536,"state":"up","type":"loopback"}},"pid":2028,"processes":18}}

Name: second
Remote: unix:/var/lib/lxd/unix.socket
Architecture: x86_64
Created: 2017/06/30 03:37 UTC
Status: Running
Type: persistent
Profiles: default
Pid: 2028
Ips:
  eth0: inet    10.50.200.19    vethVFRKLF
  eth0: inet6   fe80::216:3eff:fe63:c30b        vethVFRKLF
  lo:   inet    127.0.0.1
  lo:   inet6   ::1
Resources:
  Processes: 18
  Memory usage:
    Memory (current): 97.47MB
    Memory (peak): 100.00MB
    Swap (current): -90963968B
  Network usage:
    lo:
      Bytes received: 9.08kB
      Bytes sent: 9.08kB
      Packets received: 120
      Packets sent: 120
    eth0:
      Bytes received: 293.67kB
      Bytes sent: 5.09kB
      Packets received: 1708
      Packets sent: 56
DBUG[07-02|21:54:11] Raw response: {"type":"sync","status":"Success","status_code":200,"operation":"","error_code":0,"error":"","metadata":[]}
```

**Comment by stgraber** [comment]
Thanks. Looking at the code, it seems like it's mostly just bad error handling.

Basically what seems to be happening here is that we're getting some kind of error from the kernel and rather than fail, we're just setting the value to -1. The problem is that the swap value is then computed as "-1 - <mem>" leading to swap being reported as the negative of the memory - 1 when an error occurs.

**Comment by stgraber** [comment]
Looks to me like a better behavior would be to set to 0 so that the client doesn't show the field and it's more obvious that the value couldn't be retrieved at teh time.

**Comment by stgraber** [comment]
obviously not having the kernel tell us to go away would be nice too, but we're dealing with what we have :)

**Comment by zerkms** [comment]
Yep, sounds reasonably.

And could you please share something about the peak swap usage? As you may see it never appears in the output at all.

**Comment by stgraber** [comment]
Looking at that now

**Comment by stgraber** [comment]
So the peak reporting actually looks like a kernel problem...

The memsw values are always supposed to be the combined value of memory+swap. So to get the swap value you need to grab the memory value and substract it from the memsw value which gives you the swap.

But that doesn't seem to work for max_usage_in_bytes (peak) here:
```
root@castiana:~/data/code/lxc/lxd (stgraber/master)# cat /sys/fs/cgroup/memory/lxc/ups-monitor/memory.memsw.max_usage_in_bytes 
284143616

root@castiana:~/data/code/lxc/lxd (stgraber/master)# cat /sys/fs/cgroup/memory/lxc/ups-monitor/memory.max_usage_in_bytes 
284143616
```

Actually the values above may make sense, it's just the kernel reporting is suboptimal in this case...
Basically if your container's peak memory consumption was without the use of the swap, then there's no way for us to figure out peak values of swap usage when memory usage for the container at that time is below that pure-memory peak...

**Comment by stgraber** [comment]
just looked and there doesn't appear to be any source for that data that we could use. There's no peak value recorded in memory.stat either, just current values...

**Comment by stgraber** [comment]
So that's a bit annoying because it means that the peak swap will only ever show up in rather limited cases... And with cgroupv1 being effectively feature frozen as it's being replaced with cgroupv2, there's little chance we can improve that at the kernel level until the switch to groupv2.

**Comment by zerkms** [comment]
Thanks
PS: that's amazingly quick response and a bugfix
